SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp

```diff
@@ -11,46 +11,139 @@ namespace
+		UE_LOG(LogSpatialCommandUtils, Verbose, TEXT("Trying to start spatial service with exposed runtime ip: %s"), *RuntimeIP);
 	}
-	return FPlatformProcess::CreateProc(*SpatialGDKServicesConstants::SpatialExe, *Arguments, bLaunchDetached, bLaunchHidden, bLaunchReallyHidden, OutProcessID, PriorityModifier, OptionalWorkingDirectory, PipeWriteChild, PipeReadChild);
+
+	FSpatialGDKServicesModule::ExecuteAndReadOutput(*SpatialGDKServicesConstants::SpatialExe, Command, DirectoryToRun, OutResult, OutExitCode);
+
+	bool bSuccess = OutExitCode == 0;
+	if (!bSuccess)
+	{
+		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial start service failed. Error Code: %d, Error Message: %s"), OutExitCode, *OutResult);
+	}
+
+	return bSuccess;
 }
 
-bool SpatialCommandUtils::AttemptSpatialAuth(bool bIsRunningInChina)
+bool SpatialCommandUtils::StopSpatialService(bool bIsRunningInChina, const FString& DirectoryToRun, FString& OutResult, int32& OutExitCode)
 {
-	FString SpatialInfoResult;
-	FString StdErr;
-	int32 ExitCode;
-	ExecuteSpatialCommand(TEXT("auth login"), &ExitCode, &SpatialInfoResult, &StdErr, bIsRunningInChina);
+	FString Command = TEXT("service stop");
 
-	bool bSuccess = ExitCode == 0;
+	if (bIsRunningInChina)
+	{
+		Command += ChinaEnvironmentArgument;
+	}
+
+	FSpatialGDKServicesModule::ExecuteAndReadOutput(*SpatialGDKServicesConstants::SpatialExe, Command, DirectoryToRun, OutResult, OutExitCode);
+
+	bool bSuccess = OutExitCode == 0;
+	if (!bSuccess)
+	{
+		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial stop service failed. Error Code: %d, Error Message: %s"), OutExitCode, *OutResult);
+	}
+
+	return bSuccess;
+}
+
+bool SpatialCommandUtils::BuildWorkerConfig(bool bIsRunningInChina, const FString& DirectoryToRun, FString& OutResult, int32& OutExitCode)
+{
+	FString Command = TEXT("worker build build-config");
+
+	if (bIsRunningInChina)
+	{
+		Command += ChinaEnvironmentArgument;
+	}
+
+	FSpatialGDKServicesModule::ExecuteAndReadOutput(*SpatialGDKServicesConstants::SpatialExe, Command, DirectoryToRun, OutResult, OutExitCode);
+
+	bool bSuccess = OutExitCode == 0;
 	if (!bSuccess)
 	{
-		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial auth login failed. Error Code: %d, Error Message: %s"), ExitCode, *SpatialInfoResult);
+		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial build worker config failed. Error Code: %d, Error Message: %s"), OutExitCode, *OutResult);
 	}
 
 	return bSuccess;
 }
+
+FProcHandle SpatialCommandUtils::LocalWorkerReplace(FString ServicePort, FString OldWorker, FString NewWorker, bool bIsRunningInChina, bool bLaunchDetached, bool bLaunchHidden, bool bLaunchReallyHidden, uint32* OutProcessID, int32 PriorityModifier, const TCHAR* OptionalWorkingDirectory, void* PipeWriteChild, void * PipeReadChild)
+{
+	FString Command = TEXT("worker build build-config");
+
+	if (!ServicePort.IsEmpty())
+	{
+		Command.Append(FString::Printf(TEXT(" --local_service_grpc_port %s"), *ServicePort));
+	}
+
+	if (!OldWorker.IsEmpty())
```

Comment:
Are OldWorker and NewWorker required arguments for this class? Could validate the required arguments are set

Reply:
How do you mean?

Reply:
As in for LocalWorkerReplace these arguments aren't optional - should we somehow enforce them (assert/check or warning logs)? Not sure if assert/check is too aggressive

Reply:
Oh I see, an error sounds legit to me.